### PR TITLE
feat: nextFocus for iOS

### PR DESCRIPTION
**Pull request by Krisztiaan**
## Summary

I've found android's focus guidance api more powerful, and since it's already there, and possible to implement for iOS (mostly), why not give more options to devs?

## Changelog

[tvOS] [FEAT] - Add directional next focus targetability

## Test Plan

- create 3 views below each other
- target the next one on `nextFocusRight`
- observe the one below getting focus

### CODE REVIEW

File: React/Views/RCTTVView.m
Comment (dlowder-salesforce) [inline React/Views/RCTTVView.m:302]: These two lines are the same in all four setter methods, maybe they should be a single helper method that takes an NSNumber and returns the corresponding view in the registry.

**Review by dlowder-salesforce — CHANGES_REQUESTED** [review]
See inline comments.  Also, it would great if you could extend `RNTester/js/examples/TVFocusGudieViewExample.js` to test the new props.

File: React/Views/RCTTVView.m
Comment (Krisztiaan) [reply React/Views/RCTTVView.m]: Good catch

File: React/Views/RCTTVView.m
Comment (Krisztiaan) [reply React/Views/RCTTVView.m]: I wonder if there is already a simple helper for this in RN somewhere?

**Comment by Krisztiaan** [comment]
Converted to draft until I got an example in the sample app.

**Comment by Krisztiaan** [comment]
@dlowder-salesforce I'm wondering if this PR should instead be made against the main `react-native` repo?

**Comment by olafurnsig** [comment]
Isn't nextFocus primarily for TV devices, and since RN kinda supports AndroidTV, but dropped support for tvOS, nextFocus in the main repo is android-only?

**Comment by Krisztiaan** [comment]
While the comments may leave us to believe that, the android documentation lists these props under ["Support keyboard navigation"](https://developer.android.com/training/keyboard-input/navigation). With iOS / iPad already being able to handle keyboard, as well as tvOS, according to their docs, the [necessary api is available everywhere](https://developer.apple.com/documentation/uikit/uifocusenvironment/1648972-preferredfocusenvironments)

**Comment by dlowder-salesforce** [comment]
@Krisztiaan we could do this change in the main repo, but we should refactor it quite a bit first.  If iOS really has full support for all `UIFocusEnvironment` APIs, then we would probably want to take all the focus code and put it in `RCTView` instead of `RCTTVView`, and possibly make other changes as well.  My preference would be to get this change merged here first, and take on the refactor separately before attempting to submit to the core repo.

**Comment by dlowder-salesforce** [comment]
Also @Krisztiaan -- your Typescript extension (#102 ) is already merged.  Are the new `nextFocus` APIs already in `@types/react-native`?  If not, they should be added to `index.d.ts`.

**Comment by Krisztiaan** [comment]
> Also @Krisztiaan -- your Typescript extension (#102 ) is already merged. Are the new `nextFocus` APIs already in `@types/react-native`? If not, they should be added to `index.d.ts`.

Thanks, I was waiting for the other one to get merged first so I could rebase and apply the necessary changes I already have locally.

> My preference would be to get this change merged here first, and take on the refactor separately before attempting to submit to the core repo.

Sounds good, I'll have the example +types some time next week, then think about a good way to approach this.

**Comment by olafurnsig** [comment]
@Krisztiaan I'm very exited for this, any updates? 😄

**Comment by Krisztiaan** [comment]
Wow, I did leave you all hanging! I did not have the capacity the past few weeks, moving to a different project, however I'll try to wrap it up asap, it is not forgotten!

**Comment by Krisztiaan** [comment]
@olafurnsig please give it a try if possible. :)

**Comment by dlowder-salesforce** [comment]
@Krisztiaan thanks very much for this, I'll start reviewing as soon as I'm able

**Comment by olafurnsig** [comment]
I'm a bit lacking in time but I did a quick test with my current project.

Can't be 110% sure since my project is in ultra development mode, and I just tried this on it, it seemed like `tvParallaxProperties` magnification property was just plain ignored on TouchableHiglights, that is, no magnification occured.

Also we're missing the typings for the new properties, so I didn't go much further than that sorry. I'll try to test it again later this week by using the RNtester

**Comment by acerbetti** [comment]
Just testing with my app and it's working very well. 

Thanks for doing it, can't wait to see it merged in

**Comment by Krisztiaan** [comment]
@olafurnsig I've fixed the typings after doing this PR, but I'll add them before merging. Flow types should be up to date though.

I'll have to take a look at the parallax property, thanks for calling it out.

**Comment by dlowder-salesforce** [comment]
@Krisztiaan @olafurnsig I just did some testing with the new RNTester example, and the new focus APIs seem to be working fine 👍 .

Also, magnification seems to be working fine with `tvParallaxProperties`, at least for me.  If that's not the case for you, please send an example and I'll investigate.

**Review by dlowder-salesforce — APPROVED** [review]

**Comment by Krisztiaan** [comment]
Same result from me, @olafurnsig an example would be superb.

I'll add the types for this, and if there are no objections, we're good to go :)

**Comment by dlowder-salesforce** [comment]
Thanks @Krisztiaan ... it looks like you added the types, are we ok to merge?

**Comment by nriccar** [comment]
please merge!

**Comment by douglowder** [comment]
@nriccar this has already been merged in https://github.com/react-native-tvos/react-native-tvos/commit/abb6ee1d86b84b155c787ac45aaebdf6b4ab7245

**Comment by nriccar** [comment]
Thank you for the quick response. 

What I don't - I did investigate and read, sorry if its pretty obvious - is what the number stands for. Naturally I tried to pass a ref to the prop nextFocusDown - on every direction.

Am I missing something?

Thanks

**Comment by Krisztiaan** [comment]
Check out the example. It's not a random number, it should be a reference to another view.

**Pull request merged**
